src/devices/bus/hp9825_optroms/hp9825_optrom.cpp

```diff
@@ -0,0 +1,184 @@
+// license:BSD-3-Clause
+// copyright-holders: F. Ulivi
+/*********************************************************************
+
+    hp9825_optrom.cpp
+
+    Optional ROMs for HP9825 systems
+
+*********************************************************************/
+
+#include "emu.h"
+#include "hp9825_optrom.h"
+#include "softlist.h"
+
+// Debugging
+//#define VERBOSE 1
+#include "logmacro.h"
+
+DEFINE_DEVICE_TYPE(HP9825_OPTROM_CART, hp9825_optrom_cart_device, "hp9825_optrom_cart", "HP9825 optional ROM cartridge")
+DEFINE_DEVICE_TYPE(HP9825_OPTROM_SLOT, hp9825_optrom_slot_device, "hp9825_optrom_slot", "HP9825 optional ROM slot")
+
+struct optrom_region_t {
+	offs_t m_start;
+	offs_t m_size;
+	const char *m_tag;
+};
+
+constexpr std::array<struct optrom_region_t , 8> region_tab =
+	{
+	 0x3000 , 0x400 , "rom3000" ,
+	 0x3400 , 0x400 , "rom3400" ,
+	 0x3800 , 0x400 , "rom3800" ,
+	 0x3c00 , 0x400 , "rom3c00" ,
+	 0x4000 , 0x400 , "rom4000" ,
+	 0x4400 , 0x800 , "rom4400" ,
+	 0x4c00 , 0x400 , "rom4c00" ,
+	 0x5c00 ,0x2000 , "rom5c00"
+	};
+
+// +-------------------------+
+// |hp9825_optrom_cart_device|
+// +-------------------------+
+hp9825_optrom_cart_device::hp9825_optrom_cart_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: hp9825_optrom_cart_device(mconfig , HP9825_OPTROM_CART , tag , owner , clock)
+{
+}
+
+hp9825_optrom_cart_device::hp9825_optrom_cart_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, uint32_t clock)
+	: device_t(mconfig, type, tag, owner, clock)
+	, device_slot_card_interface(mconfig, *this)
+{
+}
+
+// +-------------------------+
+// |hp9825_optrom_slot_device|
+// +-------------------------+
+hp9825_optrom_slot_device::hp9825_optrom_slot_device(machine_config const &mconfig, char const *tag, device_t *owner)
+	: hp9825_optrom_slot_device(mconfig, tag, owner, (uint32_t)0)
+{
+	option_reset();
+	option_add_internal("rom", HP9825_OPTROM_CART);
+	set_default_option(nullptr);
+	set_fixed(false);
+}
```

Comment:
This is going to be a nasty surprise for anyone using the device.  It gets no options configured if you explicitly provide a clock frequency, but it gets an option configured if you don't explicitly provide a clock frequency.  That's not intuitive.